chain/bitcoind_events.go

```diff
@@ -339,3 +377,329 @@ func (b *bitcoindZMQEvents) txEventHandler() {
+			for i := height + 1; i <= info.Blocks; i++ {
+				newHash, err := b.client.GetBlockHash(int64(i))
+				if err != nil {
+					log.Errorf("Unable to retrieve "+
+						"block hash: %v", err)
+					continue
+				}
+
+				newBlock, err := b.client.GetBlock(newHash)
+				if err != nil {
+					log.Errorf("Unable to retrieve "+
+						"block: %v", err)
+					continue
+				}
+
+				// notify the client of the new block.
+				select {
+				case b.blockNtfns <- newBlock:
+				case <-b.quit:
+					return
+				}
+
+				// From our local mempool map, let's remove each
+				// of the transactions that are confirmed in
+				// this new block, since they are no longer in
+				// the mempool.
+				b.mempool.clean(newBlock.Transactions)
+
+				height++
+			}
+
+		case <-b.quit:
+			return
+		}
+	}
+}
+
+// txEventHandlerRPC is a goroutine that uses the RPC client to check the
+// mempool for new transactions.
+func (b *bitcoindRPCPollingEvents) txEventHandlerRPC() {
+	defer b.wg.Done()
+
+	log.Info("Started polling for new bitcoind transactions via RPC.")
+	ticker := time.NewTicker(b.cfg.TxPollingInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			// After each ticker interval, we poll the mempool to
+			// check for transactions we haven't seen yet.
+			txs, err := b.client.GetRawMempool()
+			if err != nil {
+				log.Errorf("Unable to retrieve mempool txs: "+
+					"%v", err)
+				continue
+			}
+
+			// Set all mempool txs to false.
+			b.mempool.unmarkAll()
+
+			// We'll scan through the most recent txs in the
+			// mempool to see whether there are new txs that we
+			// need to send to the client.
+			for _, txHash := range txs {
+				// If the transaction is already in our local
+				// mempool, then we have already sent it to the
+				// client.
+				if b.mempool.contains(*txHash) {
+					// Mark the tx as true so that we know
+					// not to remove it from our internal
+					// mempool.
+					b.mempool.mark(*txHash)
+					continue
+				}
+
+				// Grab full mempool transaction from hash.
+				tx, err := b.client.GetRawTransaction(txHash)
+				if err != nil {
+					log.Errorf("unable to fetch "+
+						"transaction %s from "+
+						"mempool: %v", txHash, err)
+					continue
+				}
+
+				// Add the transaction to our local mempool.
+				// Note that we only do this after fetching
+				// the full raw transaction from bitcoind.
+				// We do this so that if that call happens to
+				// initially fail, then we will retry it on the
+				// next interval since it is still not in our
+				// local mempool.
+				b.mempool.add(*txHash)
+
+				select {
+				case b.txNtfns <- tx.MsgTx():
+				case <-b.quit:
+					return
+				}
+			}
+
+			// Now, we clear our internal mempool of any unmarked
+			// transactions. These are all the transactions that
+			// we still have in the mempool but that were not
+			// returned in the latest GetRawMempool query.
+			b.mempool.deleteUnmarked()
```

Comment:
Ah ok, so we're essentially computing the set difference as we scan through mempool, then removing those items at the very end.